packages/fullscreen/src/index.ts

```diff
@@ -0,0 +1,154 @@
+      "fullscreenElement",
+      "fullscreenEnabled",
+      "fullscreenchange",
+      "fullscreenerror"
+    ],
+    // New WebKit
+    [
+      "webkitRequestFullscreen",
+      "webkitExitFullscreen",
+      "webkitFullscreenElement",
+      "webkitFullscreenEnabled",
+      "webkitfullscreenchange",
+      "webkitfullscreenerror"
+    ],
+    // Old WebKit
+    [
+      "webkitRequestFullScreen",
+      "webkitCancelFullScreen",
+      "webkitCurrentFullScreenElement",
+      "webkitCancelFullScreen",
+      "webkitfullscreenchange",
+      "webkitfullscreenerror"
+    ],
+    [
+      "mozRequestFullScreen",
+      "mozCancelFullScreen",
+      "mozFullScreenElement",
+      "mozFullScreenEnabled",
+      "mozfullscreenchange",
+      "mozfullscreenerror"
+    ],
+    [
+      "msRequestFullscreen",
+      "msExitFullscreen",
+      "msFullscreenElement",
+      "msFullscreenEnabled",
+      "MSFullscreenChange",
+      "MSFullscreenError"
+    ]
+  ];
+
+  let i = 0;
+  const l = fnMap.length;
+  const ret = {} as NormalizedFullscreenApi;
+
+  for (; i < l; i++) {
+    val = fnMap[i];
+    if (val && val[1] in document) {
+      for (i = 0; i < val.length; i++) {
+        ret[fnMap[0][i]] = val[i];
+      }
+      return ret;
+    }
+  }
+};
+
+const useFullscreen = () => {
+  const [isFullscreen, setIsFullscreen] = useState(false);
+  const [element, setElement] = useState(null);
+  const fn = getBrowserFunctions();
+
+  const eventNameMap = {
+    change: fn.fullscreenchange,
+    error: fn.fullscreenerror
+  };
+
+  const request = (element?: HTMLElement) =>
+    new Promise((resolve, reject) => {
+      const onFullScreenEntered = () => {
+        off("change", onFullScreenEntered);
+        resolve();
+      };
+
+      on("change", onFullScreenEntered);
+
+      element = element || document.documentElement;
+      setIsFullscreen(true);
+      setElement(element);
+
+      Promise.resolve(element[fn.requestFullscreen]()).catch(reject);
+    });
+
+  const on = (event: string, callback: EventCallback) => {
+    const eventName = eventNameMap[event];
+    if (eventName) {
+      document.addEventListener(eventName, callback, false);
+    }
+  };
+  const off = (event: string, callback: EventCallback) => {
+    const eventName = eventNameMap[event];
+    if (eventName) {
+      document.removeEventListener(eventName, callback, false);
+    }
+  };
+  const exit = () =>
+    new Promise(function(resolve, reject) {
+      if (!Boolean(document[fn.fullscreenElement])) {
+        resolve();
+        return;
+      }
+
+      const onFullScreenExit = () => {
+        off("change", onFullScreenExit);
+        resolve();
+      };
+
+      on("change", onFullScreenExit);
+      setIsFullscreen(false);
+      setElement(null);
+
+      Promise.resolve(document[fn.exitFullscreen]()).catch(reject);
+    });
+
+  const toggle = (element?: HTMLElement) =>
+    Boolean(document[fn.fullscreenElement]) ? exit() : request(element);
+
+  return {
```

Comment:
Is there anyway we can make the return value an array with as few elements as possible?

Reply:
Sure, each return value has a use, I assume `toggle` would make it in the most valuable but what about others?

Reply:
I started the refactor and I feel like the API is not as clean for example, I want one component to toggle fullscreen
`const { toggle } = useFullscreen();`

then I want another to react to it
`const {isFullscreen, onchange} = useFullscreen();`

with array API it might look something like this.

`const [toggle] = useFullscreen()`
then
`const [_toggle, isFullscreen, _isEnabled, onchange] = useFullscreen();`

or worse I need the fullscreen element
`const [_toggle, _isFullscreen, _isEnabled, _onchange, _onerror, _request, _exit, element] = useFullscreen();`

Reply:
let me know your thoughts.

Reply:
cc @simbathesailor @maciekgrzybek

Reply:
I think that all depends on what we want to expose to the API? Do we really need to use literally everything? If so, I don't think that returning that object would be a problem.

Reply:
For example, why do we need to expose the element? Shouldn't that be passed as an argument?

Reply:
It is an optional argument if you wanted to make the entire app fullscreen you wouldn't pass request or toggle any arguments this will by default target the root of the application.

If you have multiple things that can go full screen it might be helpful to be able to use this property to know what is fullscreen at the moment.

Reply:
Hmm, but shouldn't it be a different 'instance' of hook called for each element?

Reply:
So there will be always only one element (or root) per hook.

Reply:
Good point, I updated it so it will derive the default hook state from document. This will allow the hook to share state across the application in a sense

Component 1: Set Document to full screen
Component 2: Mounts with default state derived from `document`

thoughts on this?